### PR TITLE
[remove nixpkgs] part 1: write info for nix store paths in lockfile

### PR DESCRIPTION
**Pull request by savil**
## Summary

**Motivation For "Remove Nixpkgs"**
Today, when adding a versioned package, Devbox will look up the versioned package’s metadata by mapping the user-specified version to a nixpkgs commit hash and an attribute path. This `nixpkgs/<commit hash>#attributePath` is used in the generated flake.nix. 

We want to remove the dependency on the nixpkgs commit hash, so that the user need not download that nixpkgs artifact, and spend time resolving it.

**Essence of the approach**
Today, when adding a versioned package, Devbox will look up the versioned package’s metadata by mapping the user-specified version to a nixpkgs commit hash and an attribute path. This `nixpkgs/<commit hash>#attributePath` is used in the generated flake.nix. 
```
builtins.fetchClosure {
  fromStore = "https://cache.nixos.org";
  fromPath = /nix/store/r2jd6ygnmirm2g803mksqqjm4y39yi6i-git-2.33.1;
  toPath = /nix/store/ldbhlwhh39wha58rm61bkiiwm6j7211j-git-2.33.1;
}
```

if we can map `git@2.33.1` to the `/nix/store` paths, then we need not download nixpkgs.

**Implementation in this PR**

This PR focusses on the resolving packages to the "locked" representation in `devbox.lock`. Anytime a package is added or removed, we check if the `devbox.lock` needs additional `SystemInfo` for each package. This `SystemInfo` can be used to derive `fromPath` and `toPath` in `builtins.fetchClosure`.

Next PR will read this information from the lockfile to generate a `flake.nix` that uses `builtins.fetchClosure`.

cc @mikeland73 

## How was it tested?

Using the devbox repository, did `devbox add hello`.

```
diff --git a/devbox.json b/devbox.json
index 7b263cb4..55ee3f8a 100644
--- a/devbox.json
+++ b/devbox.json
@@ -2,7 +2,8 @@
   "packages": [
     "go@1.20",
     "actionlint@1.6.23",
-    "golangci-lint@1.52.2"
+    "golangci-lint@1.52.2",
+    "hello@latest"
   ],
   "env": {
     "PATH": "$PATH:$PWD/dist"
diff --git a/devbox.lock b/devbox.lock
index 7dffa054..3baae253 100644
--- a/devbox.lock
+++ b/devbox.lock
@@ -4,17 +4,78 @@
     "actionlint@1.6.23": {
       "last_modified": "2023-03-31T22:52:29Z",
       "resolved": "github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#actionlint",
-      "version": "1.6.23"
+      "version": "1.6.23",
+      "systems": {
+        "aarch64-darwin": {
+          "System": "aarch64-darwin",
+          "from_hash": "zx4pz0wxyh7fww34fmkvy11rjx36jn14",
+          "store_name": "actionlint",
+          "store_version": "1.6.23"
+        },
+        "x86_64-linux": {
+          "System": "x86_64-linux",
+          "from_hash": "cffgrw9ynyfxirqpakx96m4v080pnmxz",
+          "store_name": "actionlint",
+          "store_version": "1.6.23"
+        }
+      }
     },
     "go@1.20": {
       "last_modified": "2023-05-01T16:53:22Z",
       "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#go",
-      "version": "1.20.3"
+      "version": "1.20.3",
+      "systems": {
+        "aarch64-darwin": {
+          "System": "aarch64-darwin",
+          "from_hash": "pn1m11aq5ridxm6qcvpazmrrj3kqln71",
+          "store_name": "go",
+          "store_version": "1.20.3"
+        },
+        "x86_64-linux": {
+          "System": "x86_64-linux",
+          "from_hash": "mzrw2zphz9xx9s3qq8x5zpf18isp0jga",
+          "store_name": "go",
+          "store_version": "1.20.3"
+        }
+      }
     },
     "golangci-lint@1.52.2": {
       "last_modified": "2023-05-01T16:53:22Z",
       "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#golangci-lint",
-      "version": "1.52.2"
+      "version": "1.52.2",
+      "systems": {
+        "aarch64-darwin": {
+          "System": "aarch64-darwin",
+          "from_hash": "rp380cz8gxv8pw5zq985yblym19vr9g8",
+          "store_name": "golangci-lint",
+          "store_version": "1.52.2"
+        },
+        "x86_64-linux": {
+          "System": "x86_64-linux",
+          "from_hash": "9cl7cbahf1zqfwc3mfp44bkd9h82c1pg",
+          "store_name": "golangci-lint",
+          "store_version": "1.52.2"
+        }
+      }
+    },
+    "hello@latest": {
+      "last_modified": "2023-05-01T16:53:22Z",
+      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#hello",
+      "version": "2.12.1",
+      "systems": {
+        "aarch64-darwin": {
+          "System": "aarch64-darwin",
+          "from_hash": "qh8s6yb93b1wl0jxg5i0pfq2rf7qm44j",
+          "store_name": "hello",
+          "store_version": "2.12.1"
+        },
+        "x86_64-linux": {
+          "System": "x86_64-linux",
+          "from_hash": "8n3pr83b7aab975gwrd6hnk10pqphr59",
+          "store_name": "hello",
+          "store_version": "2.12.1"
+        }
+      }
     }
   }
 }
\ No newline at end of file

```

### CODE REVIEW

**Comment by savil** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1208** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/1208" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jetpack-io/devbox/1208?utm_source=stack-comment).

File: internal/searcher/client.go [hunk 1]
```diff
@@ -66,17 +68,68 @@ func (c *client) Resolve(pkg string) (*lock.Package, error) {
...
+	}
+
+	userSystem, err := nix.System()
+	if err != nil {
+		return nil, err
+	}
+
+	var systemInfo *lock.SystemInfo
+	for sysName, sysInfo := range result.Systems {
+		if sysName == userSystem {
```
Comment (gcurtis) [inline internal/searcher/client.go:120]: Should we just add all of the systems we get back? There should be at most 4. It would make moving between local and cloud faster if your local system isn't x86 Linux.

**Review by gcurtis — APPROVED** [review]

File: internal/searcher/client.go (hunk 1 above)
Comment (savil) [reply internal/searcher/client.go:120]: Actually, yes, we should. Good point.

It'll also simplify some of the "merge" logic I have elsewhere in this PR.

File: internal/impl/devbox.go [hunk 2]
```diff
@@ -88,7 +89,17 @@ func Open(opts *devopt.Opts) (*Devbox, error) {
...
+	// TODO savil: this is bad for perf, and so remove before enabling feature.
+	// this hack is to workaround an import cycle: lock -> nix -> lock
+	userSystem := ""
+	if featureflag.RemoveNixpkgs.Enabled() {
+		userSystem, err = nix.System()
+		if err != nil {
+			return nil, err
+		}
+	}
+	lock, err := lock.GetFile(box, searcher.Client(), userSystem)
```
Comment (mikeland73) [inline internal/impl/devbox.go:102]: Use interfaces to get around dependency cycles. For example if you add `devbox.System()` and update the interface in `lock` pkg you are all good.

File: internal/lock/lockfile.go [hunk 3]
```diff
@@ -24,24 +23,41 @@ type File struct {
 	devboxProject
 	resolver
 
-	LockFileVersion string              `json:"lockfile_version"`
-	Packages        map[string]*Package `json:"packages"`
+	LockFileVersion string `json:"lockfile_version"`
+
+	// Packages is keyed by "canonicalName@version"
```
Comment (mikeland73) [inline internal/lock/lockfile.go:28]: This is not strictly true. packages are keyed by any value that can go in `packages` in devbox.json

File: internal/lock/lockfile.go [hunk 4]
```diff
@@ -24,24 +23,41 @@ type File struct {
...
+	Packages map[string]*Package `json:"packages"`
+
+	system string
```
Comment (mikeland73) [inline internal/lock/lockfile.go:31]: Why is this here? it's not used

File: internal/lock/lockfile.go (hunk 3 above)
Comment (savil) [reply internal/lock/lockfile.go:28]: isn't that always with `@latest` or `@<concrete-version>` ?

File: internal/impl/devbox.go (hunk 2 above)
Comment (savil) [reply internal/impl/devbox.go:102]: Ugh, no, please lets not add more cruft to the `impl` package. 

Instead, would prefer simpifying `nix` package so it is not dependent on `lockfile`. From inspection, it the `profile` functions in `nix` have a lot of devbox-specific logic and so should be moved out into a different package.
